src/main/java/io/cdap/plugin/gcp/datastore/source/DatastoreRecordReader.java

```diff
@@ -67,14 +73,30 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
                         .setNamespaceId(config.get(DatastoreSourceConstants.CONFIG_NAMESPACE))
                         .setProjectId(config.get(DatastoreSourceConstants.CONFIG_PROJECT)))
       .build();
+
+    entityResultList = new ArrayList<>();
+    createEntityResultsList(query, request);
+    taskAttemptContext.getCounter(FileInputFormatCounter.BYTES_READ).increment(batchSerializedSize);
+    results = entityResultList.iterator();
+    index = 0;
+  }
+
+  private void createEntityResultsList(Query query, RunQueryRequest request) throws IOException {
     try {
       QueryResultBatch batch = datastore.runQuery(request).getBatch();
-      taskAttemptContext.getCounter(FileInputFormatCounter.BYTES_READ).increment(batch.getSerializedSize());
-      results = batch.getEntityResultsList().iterator();
+      batchSerializedSize += batch.getSerializedSize();
+      entityResultList.addAll(batch.getEntityResultsList());
+      LOG.debug("Batch moreResultsType: {}", batch.getMoreResults());
+      // Datastore API only returns up to 300 items. Need to use cursor pagination if there is more results
+      if (batch.getMoreResults() != QueryResultBatch.MoreResultsType.NO_MORE_RESULTS) {
+        ByteString cursor = batch.getEndCursor();
+        query = query.toBuilder().setStartCursor(cursor).build();
+        request = request.toBuilder().setQuery(query).build();
+        createEntityResultsList(query, request);
```

Comment:
Please don't use recursion instead of loop. You can easily run out of stack

Reply:
Also, does it mean we need to read the whole split before serving it? It does not seem optimal

Reply:
I think so. It seems like the API only supports setting cursors at the query level (and not the request/response level)

Reply:
And that's fine. Just don't use recursion a run next batch in "nextKeyValue" call, not "initialize" call. This would ensure we don't accumulate data in memory.
Also don't accumulate the counter, but call `increment` as needed